Support pointer-to-member-function in def_buffer

bmerry opened the pull request:
Closes #857. Instead of calling the function object directly, it wraps
it in std::bind and then immediately calls that. This is a poor-man's
form of std::invoke (not available in C++11 or C++14).

dean0x7d commented:
The inclusion of `<functional>` in the main `pybind11.h` header may be a compilation time issue. At least I assume so, given that `pybind11/functional.h` is an optional extra and not part of the core headers. cc @wjakob

jagerman commented:
The use of `<functional>` is probably a non-starter (not including the rather heavy-weight `<functional>` in core pybind was a design decision from way back).

An overload that accepts a member function pointer and converts it into lambda, on the other hand, could work, and shouldn't be noticeably more code.

Also, `tests_buffers` seems like a better place than `test_issues`.  The latter is quasi-deprecated, with a vague long-term plan to move everything out of it into more appropriate places, and to at least not add anything more into it.

bmerry commented:
Ok, I'll give it another go. I guess the simplest thing then would be to provide an overload of `def_buffer` that takes any PTMF, wraps it in a lambda, and passes it to the original overload?

bmerry commented:
I've reworked it and changed the PR title to reflect the new version. I haven't flattened the history at this point.

Do you think it is also worth adding an overload for pointer to data member (so that the class can initialise a buffer_info member when constructed and then we just pass a pointer to that member to def_buffer?

bmerry commented:
The pypy build is crashing. Is that expected due to https://bitbucket.org/pypy/pypy/issues/2444 (the other to_python test in the file is disabled because of that, but I don't want to do the same here until someone can confirm that I'm not just hiding a real bug, particularly since that pypy bug seems into indicate a leak rather than a double-free.

dean0x7d commented on tests/test_buffers.cpp:
```diff
@@ -74,6 +74,34 @@ class Matrix {
 };
 
+struct PTMFBuffer
+{
```
Brace style: same line.

dean0x7d commented on tests/test_buffers.cpp:
```diff
@@ -74,6 +74,34 @@ class Matrix {
+};
+
+class ConstPTMFBuffer
+{
```
Same.

bmerry commented:
> I've just noted a minor brace style issue.

Fixed, thanks.

jagerman commented:
Looks good to me.

jagerman commented:
Can you squash the changes and update the commit messages into a single commit?

bmerry commented:
Sure. Should I do that as a force-push to the branch?

jagerman commented:
Yup.

bmerry commented:
Done, and also added the marker to disable the new test in PyPy. Once Travis and Appveyor are happy I think it should be good to go.

jagerman commented:
Looks like the last appveyor test got stuck, but all the rest passed so this is good to go.
